introduce emqx_conf application. (#50)
emqx/eip: active/0016-emqx-conf.md

# EMQX Configuration Manager

## Changelog

* 2020-10-21: @zhongwencool init draft

## Abstract

Introduces how to  initialize/hot update/reset configurations.

## Motivation

Try to find a suitable way to:

- Hot update configurations via HTTP API.
- Maintain data consistency for the same configuration, keeping the same configuration in the cluster must be the same, such as `authorization.source`, rules engine's rules.
- Different nodes of the cluster may set different values for the same configuration item, such as: use different `rate_limit` settings for machines with different hardware limitations.

## Design

### Configuration files

#### emqx.conf

> Emqx reads `emqx.conf` forconverting this hocon file into Erlang format at startup, and `emqx.conf` has only 2 lines by default.

Reviewer: My concern about "only 2 lines"
it is probably going to be a quite large task.
I would recommend NOT to scope it in this EIP.

We can come back and update this one as step 2,
or create another EIP to focus on that change.

> Emqx reads `emqx.conf` forconverting this hocon file into Erlang format at startup, and `emqx.conf` has only 2 lines by default.

Author: Create a new EIP should be better.
I update:
...
and `emqx.conf` explicitly include `override.conf` at the end of the file.
....
For now, we need to integrate all the configurations into `emqx.conf`. But it is planned that `emqx.conf` will no longer contain configurations that use default values in the next phase(propose another EIP), the default is only embedded in the code. Also, users can view all configuration items via the HTTP API (described later). This would have 2 benefits: 

- In subsequent version upgrades, adding/removing/updating configurations will not be overwritten by `emqx.conf`. 
- This allows the user to focus only on the configuration that has been modified.

```erlang
include "./data/emqx_cluster_override.conf"

> include "./data/emqx_local_override.conf"

Author: https://github.com/emqx/hocon/pull/147

- The node will copy this file from the longest surviving core node before initializing the configuration, , this file will be added to initialize the configuration together.
- When the node is updated, the configuration within the cluster is updated via cluster call and persisted to this file.
- This file must be kept the same for all nodes, we will add an extra process to check the content of the file periodically, and alarm after 3 continuous differences are found.

#### emqx_local_override.conf

- This file can only be modified via the API and manual modification of file directly is not supported.

- When the configuration of a specific node is updated via HTTP API, it will be persisted to this file.

#### emqx_conf application
Before, the initialization of the configuration file, cluster_call, was done through the `emqx_machine`, we will split this part of the functionality from the` emqx_machine` and make a new application `emqx_conf`

The role of this application is to
- Converting the configuration from hocon format to erlang sys.config format at initialization.

> - Manage hot updates and deletions of configurations within the cluster.

Reviewer: ```suggestion
- Manage live-updates and deletions of the configurations, and replicate across the cluster
```

> The basic node configuration hot update remains in emqx, in other words: `emqx_config_handler`, remains unchanged, but other apps that want to update the configuration must call through the `emqx_conf`'s' API, which cannot call emqx API directly.

Reviewer: This is a new feature for 5.0, so we are comparing it to 4.3, not 5.0 beta.1
Describing the changes on beta.1 will soon become stale.
So we should introduce `emqx_config_handler` as a new module.

It should be enough just to elaborate a bit on the layering of the APIs (line 64)

The specific flow is: 

```bash
Other Apps(eg: emqx_resource) => emqx_conf => emqx API.
```

### HTTP API design

> 

Reviewer 2: Do we need a API for `validate` the config before apply?

#### Get the whole configurations.

```erlang
#{
  get => #{

>     description => <<"Get all the configurations of cluster, including hot and non-hot updatable items.">>,

Reviewer: ```suggestion
    description => <<"Get all the configurations of a given node, or all nodes in the cluster.>>,
```

    parameters => [
       {node, hoconsc:mk(typerefl:atom(),
          #{in => query, required => false, example => <<"emqx@127.0.0.1">>,

>           desc => <<"Node's name: If you have this field defaulted, the whole cluster is return.">>})},

Reviewer: ```suggestion
          desc => <<"Node  name. When this parameter is not provided, configs for all nodes in the cluster are returned">>})},
```

       {debug, hoconsc:mk(typerefl:boolean(), #{in => query, required => false,

>           desc => <<"Carries debug(meta data) information, such as:line number,default value,documentation">>})}],

Reviewer: I am not sure how easily fill in default value and documentation into metadata.
Maybe leave this feature out for now.

```suggestion
          desc => <<"Carries debug (metadata) information, such as file name and line number">>})}],
```

Author: ok, file name and line number first come.

#### Update specific configuration

```erlang
schema("/configs/:rootname") ->
    #{
      get => #{
            description => <<"Get the sub-configurations">>,
            parameters => [
              {node, hoconsc:mk(atom(), #{in => query, required => false})},
              {debug, hoconsc:mk(typerefl:boolean(), #{in => query, required => false})}
                          ],
            responses => #{
                200 => #{<<"$node">> => config_list()},
                404 => emqx_dashboard_swagger:error_codes(['NOT_FOUND'], <<"config not found">>)
            }
        },
        put => #{
            description => <<"Update the sub-configurations">>,
            parameters => [{node, hoconsc:mk(atom(), #{in => query, required => false})}],
            requestBody => config_list(),
            responses => #{
                200 => #{<<"$node">> => config_list()},
                400 => emqx_dashboard_swagger:error_codes(['UPDATE_FAILED'])
            }
        }
    }.
```

- get specific configuation, such as: `/configs/emqx_dashboard` will return :

  ```erlang
  #{'emqx@127.0.0.1' => 

>      #{default_password => "public",

Reviewer: There should be a `sensitive` flag in the schema for this field, and the value should be reported back as `"******"` in the API.

       default_username => "admin",
       listeners =>
         [#{backlog => 512,inet6 => false,ipv6_v6only => false,
            max_connections => 512,num_acceptors => 4,port => 18083,
            protocol => http,send_timeout => 5000}],
       sample_interval => 10,token_expired_time => 3600000},
    'emqx1@127.0.0.1' => ...
  }   
  ```

> - Update specific configuation without query string, will modify the configuration of all nodes in the cluster and persist it in `emqx_cluster_override.conf`. In other words, it is the configuration that keeps the cluster strongly consistent.

Reviewer: ```suggestion
- Update specific configuation without the 'node' query string will modify the configuration of all nodes in the cluster and persist it in `emqx_cluster_override.conf`.
```
